libtenzir/include/tenzir/pipeline.hpp

```diff
@@ -317,7 +318,43 @@ class operator_base {
   /// which is implied by `sink <=> OPT | sink`. If `order = schema`, this
   /// resolves to `sink <=> interleave | pass | sink`, which follows from what
   /// we may assume about `sink`.
-  virtual auto optimize(expression const& filter, event_order order) const
+
+  /// # Projection Pushdown
+  ///
+  /// The selection is a projection pushdown that is represented as a vector of
+  /// columns to select. This vector can be moved around a pipeline similarly to
+  /// an expression. The implementation must ensure that the selected columns
+  /// are preserved in the correct order throughout the pipeline processing
+  /// stages. For an otherwise unknown pipeline `sink`, the following
+  /// equivalences must hold:
+  ///
+  /// ~~~
+  /// this | select columns | sink
+  /// <=> select columns | this | sink
```

Comment:
**Does this sound better (italics is new part)?**


#Projection Pushdown

The selection is a projection pushdown represented as a vector of
columns to select. This vector can be moved around a pipeline similar to an
expression. The implementation must ensure that the selected columns are
preserved in the correct order throughout the pipeline processing stages. For an
otherwise unknown pipeline `sink,` the following equivalences must hold :
  ~~~
  this  | select columns | sink <=> select columns | this | sink
  ~~~

_The input columns will differ from the output columns(i.e., if columns 0 and 4 are
selected, then 1, 2, and 3 remain), but the
individual parse operator must handle this functionality. Optimize allows a modified parser with specified
columns to be returned._

If the operator modifies the schema of the events, the
selection must be adjusted accordingly to ensure that the specified columns are
correctly mapped and preserved in the resulting schema. _The pipeline
implementation can block the upward movement of the select operator by
passing a std::nullopt into the optimized result. Semantically, an empty vector
of selection means no selection was provided, whereas a nullopt selection means
the prior selection is being blocked._

#Example

For an operator that filters and then selects columns,
the optimized output must ensure that the filter is applied before the
selection to maintain the correct semantics :

  ~~~
  where expr | OPT | select columns | sink <==> where expr | select columns |
OPT | sink
  ~~~

  This ensures that filtering and selection are treated as distinct operations,
  and the correct behavior is preserved in the optimized pipeline.

Reply:
It would be best if you could write the exact requirements down succinctly. It might be easiest to do this semi-formally, as we have done for predicate pushdown.

Let me try my luck, and you tell me whether what I wrote captures what you are doing. I did not take a close look at the implementation yet, so what I write might also be incorrect. Just let me know. I will limit myself to the selection optimization. The full interaction between predicate, selection and ordering optimizations also needs to be described somewhere, but this is a starting point (please do not use it directly).

---

If `opt.selection` has a value (where `opt` is the returned `optimize_result`), it must hold that

```
this | select selection | sink 
<=> select opt.selection | opt.replacement | sink
``` 

Otherwise: 

```
this | select columns | sink 
<=> opt.replacement | select columns | sink
```

If `opt.replacement` is `nullptr`, then we interpret it as `pass` above.  

---

Is this correct? 

If so, can you extend this to describe how it interacts with the predicate optimization? In particular, I am interested the order of `select` and `where` in the requirements.

Reply:
Thoughts?

If `opt.selection` has a value (where `opt` is the returned `optimize_result`), it must hold that

```
this | select columns | sink 
<=> select opt.selection | this | sink
``` 

If `opt.selection` has no value and the `current_selection`, representing the last select seen call, is not empty, it must hold that
```
this | sink 
<=> this | select columns | sink
```
Here, the previous `select` is inserted as a block. In all other cases where`opt.selection` has no value, it must hold that no `select` operator is being pushed up the pipeline. 

---

If `opt.filter` and `opt.selection` can never contain values simultaneously. They must push up their optimization information, or the filter/selection information will never be recorded.

Note: An alternative to not allowing both filter and select to be pushed together up the pipeline is to force this combination to have one semantic meaning, i.e., `opt.filter` and `opt.selection` means `select | this | where filter | sink`. 

---

`opt.replacement` operates in isolation from `opt.selection ` and `opt.filter`. For example, in the do_not_optimize() function, opt.replacement() is a copy of the original operator.

```
this | select selection | sink 
<=> select opt.selection | opt.replacement | sink
``` 

If `opt.replacement` is `nullptr`, then it must hold that no changes were made to the original operator in the pipeline.